protocols/request-response/src/lib.rs

```diff
@@ -677,7 +677,24 @@ where
         }
     }
 
-    fn on_dial_failure(&mut self, DialFailure { peer_id, .. }: DialFailure) {
+    // Removes the connection for a peer if it exists.
+    fn remove_connection_for_peer(&mut self, peer: PeerId, connection_id: ConnectionId) {
+        if let Some(connections) = self.connected.get_mut(&peer) {
+            connections
+                .iter()
+                .position(|c| c.id == connection_id)
+                .map(|p: usize| connections.remove(p));
+        }
+    }
```

Comment:
`ConnectionId`s are unique even across peers, meaning you should be able to remove a connection just by searching for the `ConnectionId`, regardless of the `PeerId`.